16972 add assignee responsible to wp form

hschink opened the pull request:
[`* `#16972` Support assignee/responsible property on work package forms and PATCH endpoint`](https://community.openproject.org/work_packages/16972)

Includes
- https://community.openproject.org/work_packages/17426
- https://community.openproject.org/work_packages/17432
- https://community.openproject.org/work_packages/17436
- [x] Path service

Commit:
Test available assignees/responsibles on project endpoint

Commit:
Move available assignees/responsibles to project endpoint

Commit:
Test responsible/assignee in WP form schema

Commit:
Add responsible/assignee to WP form schema

Commit:
Test WP form payload includes assignee/responsible

Commit:
Add assignee/responsible to WP form payload

Commit:
Remove superfluous methods

Commit:
Test API V3 WP PATCH assignee/responsible update

Commit:
Implement API V3 WP PATCH assignee/responsible update

Commit:
Test assignee/responsible update on WP form endpoint

Commit:
Fix WP validation

In contrast to my expectation the contract does not trigger the WP's validation.
Thus, those validations need to be triggered separately.

This issue became visible when removing the call to WP's valid? method removed.
The call is removed because it empties the list of errors which removes those
errors added by the contract.

TeatroIO commented:
I've prepared a stage. <a href='http://feature-16972-add-assignee-responsible-to-wp-form.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

ulferts commented on lib/api/v3/projects/projects_api.rb:
```diff
@@ -45,6 +45,28 @@ class ProjectsAPI < Grape::API
+            end
+
             mount API::V3::Categories::CategoriesAPI
             mount API::V3::Versions::VersionsAPI
```
Can me move both into a separate class? It is possible to split them from the rest  so I guess we should.

ulferts commented on lib/api/v3/work_packages/form/work_package_attribute_links_representer.rb:
```diff
@@ -52,6 +52,42 @@ class WorkPackageAttributeLinksRepresenter < Roar::Decorator
+                   getter: -> (*) {
+                     id = represented.assigned_to_id
+
+                     { href: id ? "#{root_path}api/v3/users/#{id}" : nil }
```
We need to start introducing some path service. Otherwise, changing a path will become very hard.

ulferts replied:
Why do you use that condition within the getter? Why not reuse the pattern from above (https://github.com/opf/openproject/pull/2239/files#diff-da421ae54f2201015bc50791b17585efR54).

Will the link be displayed in either case?

hschink replied:
The `if` statement will prohibit the rendering of this property. But we want to have it even if no assignee/responsible is set.

The `if` in line 54 is plain wrong: I'll remove it.

ulferts commented on lib/api/v3/projects/projects_api.rb:
```diff
@@ -45,6 +45,28 @@ class ProjectsAPI < Grape::API
+            resource :available_assignees do
+
+              get do
+                authorize(:view_project, context: @project)
```
The necessary permission for this is not specified @NobodysNightmare. I would opt for :view_work_packages as this is the intended audience. :view_project has the additional disadvantage that it is a public permission meaning everybody having any permission within the project also has this one.

NobodysNightmare replied:
I'll specify it within this PR...

hope that is okay for you...

NobodysNightmare replied:
@ulferts I guess everyone that can view work packages is also able to see the users of a project, right? Because we basically allow the client to list (most?) users of the project...

ulferts replied:
Right, but if you want to assign the work package to somebody, you must know to whom it is possible. The question, as with all requests for users, is more with regards to the fields returned IMO.

ulferts commented on lib/api/v3/work_packages/form/form_representer.rb:
```diff
@@ -76,7 +76,7 @@ def initialize(model, options = {})
               method: :patch
             } if @current_user.allowed_to?(:edit_work_packages, represented.project) &&
-                 represented.valid?
+                 represented.errors.empty?
```
??? I don't get it. A comment please.

My understanding is, that you with this rely on somebody having called valid before, which sounds bad.

NobodysNightmare committed:
specify required permissions

[ci skip]

Commit:
Always render the WP status

The WP status is always set, thus we don't need special precautions here.

ulferts commented on lib/api/v3/work_packages/form/work_package_attribute_links_representer.rb:
```diff
@@ -52,6 +52,42 @@ class WorkPackageAttributeLinksRepresenter < Roar::Decorator
+
+            # The return value -1 will trigger a validation error. This is
+            # intended as no valid user resource was passed
+            (resource && resource[:ns] == 'users') ? resource[:id] : -1
```
Can't we just exit here with a 422 or even a 400 as the request was malformed?

ulferts commented on lib/api/v3/work_packages/work_package_contract.rb:
```diff
@@ -87,26 +95,32 @@ def readonly_attributes_unchanged
+          return if id.nil? || !model.changed.include?(id_attribute)
+
+          unless user_visible?(id, list)
+            errors.add attribute, "with id=#{id} does not exist"
```
This error message is not very good. If we can get rid of that magical -1, we could focus on what the list of available values provides us and return that the user is not within the list of possible assignees/responsibles.

hschink replied:
@ulferts which error message would you expect instead?

ulferts replied:
Do we have access to the AR (and it's connection to I18n) here? Just asking because we probably shouldn't have to many human readable strings within the code. 

We can definitely do someting like `I18n.t('activerecord.errors.messages.inclusion', locale: :en)` although it might make sense to take another look at what reform does with regards to errors and their localisation.

The `inclusion` might not offer the best translation, though. I just propose moving stuff like this to the localisation files.

The message could be something like "is not within the list of valid users".

Commit:
Test response for invalid resource link

Commit:
Throw 422 if invalid resource link is passed

ulferts commented on lib/api/v3/work_packages/work_package_contract.rb:
```diff
@@ -87,26 +95,32 @@ def readonly_attributes_unchanged
         end
 
-        def error_message(path)
-          I18n.t("activerecord.errors.models.work_package.attributes.#{path}")
```
Now I see this. :+1:

ulferts commented on lib/api/errors/form/invalid_resource_link.rb:
```diff
@@ -0,0 +1,45 @@
+          message = "For property #{property} a resource of type #{expected}" \
+                    " is expected but got a resource of type #{actual}"
+
+          super(message)
```
Looks good :+1:

ulferts commented:
The path service does not necessarily have to be part of this PR. It's just something that occurred to me while reading the PR.

Commit:
Remove superfluous API V3 models

Commit:
Add path helper for API V3

Commit:
Adapt specs to path helper usage

Commit:
Use API V3's path helper

Commit:
Remove obsolete link to available statuses

This was replaced by the work package form

ulferts commented on lib/api/v3/work_packages/work_package_representer.rb:
```diff
@@ -107,43 +107,35 @@ def initialize(model, options = {}, *expand)
           } unless represented.assigned_to.nil?
         end
 
-        link :availableStatuses do
```
@hschink the documentation does still mention it.

hschink replied:
@ulferts fixed in https://github.com/opf/openproject/commit/9c33c168f4d9ad277fc40c4793bd15ba69c95f6e

ulferts commented on lib/api/v3/utilities/path_helper.rb:
```diff
@@ -0,0 +1,152 @@
+
+        def api_v3_work_package_watchers_path(id)
+          "#{api_v3_work_package_path(id)}/watchers"
+        end
```
All of those methods have `api_v3_<unique>_path`. How about changing this so that the call is `api_v3_path.<unique>`?

hschink replied:
Do you like to see a method call here?

ulferts replied:
The method call should be made in the endpoint/representer using this. But the structure in here could change to:
<pre>

module API
  module V3
    module Utilities
      module PathHelper
        include API::Utilities::UrlHelper
          class ApiV3Path

```
        ....

        def self.work_package_relations(id)
        ....
        end

        def self.work_package_watchers(id)
        ...
        end


        ....

      end
   def api_v3_paths
     ApiV3Path
   end
end
```

end 
end

sorry about the indentation.

ulferts commented:
Looks good except for the two remarks.

ulferts commented:
@hschink, I took the liberty to rebase this.

ulferts commented:
When I POST this:

<pre>
{
"_type": "WorkPackage",
"_links": {
    "status": {
        "href": "/api/v3/statuses/3"
    },
    "assignee": {
        "href": null
    },
    "responsible": {
        "href": null
        }
    },
    "lockVersion": 6,
    "subject": "illo quo eos voluptatem numquam sint perspiciatis qui",
    "rawDescription": "Cubitum armo ancilla sto cibus amissio adsidue campana. Corrigo ea damnatio nesciunt agnosco decretum ancilla conicio. Voluntarius derelinquo textus sit molestias iusto sol. Balbus carcer cervus absque tergeo venio tondeo. Vero animi cursus defleo corrumpo ventosus."
}
</pre>

to the `/forms` endpoint I get back:

`For property status a resource of type Status is expected but got a resource of type .`

This is:
- Not a valid json response
- unexpected as the href for the status was taken from the payload provided via the first form request. It is the href to the current status.

This initial payload includes some attributes we do not support (yet) so I guess they should be removed (e.g. projectId, versionId). I had to remove them as the server would otherwise blame me for trying to update attributes that are not writeable. Below is the payload part of an initial POST to /form:

<pre>
"_type": "WorkPackage",
"_links": {
"status": {
"href": "/api/v3/statuses/3"
},
"assignee": {
"href": null
},
"responsible": {
"href": null
}
},
"lockVersion": 6,
"subject": "illo quo eos voluptatem numquam sint perspiciatis qui",
"rawDescription": "Cubitum armo ancilla sto cibus amissio adsidue campana. Corrigo ea damnatio nesciunt agnosco decretum ancilla conicio. Voluntarius derelinquo textus sit molestias iusto sol. Balbus carcer cervus absque tergeo venio tondeo. Vero animi cursus defleo corrumpo ventosus.",
"projectId": 1,
"startDate": "2014-03-07T00:00:00+00:00",
"dueDate": "2014-04-28T00:00:00+00:00",
"versionId": 14,
"createdAt": "2014-02-19T09:03:27Z",
"updatedAt": "2014-11-12T16:44:18Z"
</pre>

Otherwise a client can not simply take the payload and use it to template his next request.

Commit:
Do not render readonly attributes in WP form

The WP contract checks if readonly attributes are written. If we remove the
properties completely, the contract will not be able to detect those invalid
writes. Thus, we must still allow writes to readonly attributes.

Commit:
Test that non-existing status returns a valid form

Commit:
Do not access status object that might not exist

Commit:
Test handling of nil assignee/responsible

Commit:
Fix error message format

Commit:
Test multiple errors on form request

Commit:
Validate WP even if contract validation failed

Commit:
Map 'property_id' errors to 'property' errors

Commit:
Test allowed status are preserved on status change

Commit:
Preserve allowed statuses after status change

myabc commented on lib/api/v3/root_representer.rb:
```diff
@@ -35,23 +35,27 @@ module V3
       link 'project' do
         {
-          href: "#{root_path}api/v3/project/{project_id}",
+          href: api_v3_paths.projects + '/{project_id}',
```
@hschink why not `api_v3_paths.project(project_id)`?

myabc commented on lib/api/v3/work_packages/form/work_package_attribute_links_representer.rb:
```diff
@@ -37,21 +37,68 @@ module Form
+                   getter: -> (*) {
+                     id = represented.assigned_to_id
+
+                     { href: id ? api_v3_paths.user(id) : nil }
```
@hschink very minor, but `api_v3_paths.user(id) if id` looks more 'Ruby-ish' to me than the ternary operator here.

hschink replied:
May looks more 'Ruby-ish' but isn't syntactically correct Ruby (just tried it to be sure). :wink:

myabc replied:
(You'd need to put it in parentheses)

On Thursday, November 27, 2014, Hagen Schink notifications@github.com
wrote:

> In
> lib/api/v3/work_packages/form/work_package_attribute_links_representer.rb:
> 
> > -                     represented.status_id = resource[:id] if resource[:ns] == 'statuses'
> > -                     represented.status_id = resource[:id] if resource
> > -                   }
> >   +
> > -          property :assignee,
> > -                   exec_context: :decorator,
> > -                   getter: -> (*) {
> > -                     id = represented.assigned_to_id
> >   +
> > -                     { href: id ? api_v3_paths.user(id) : nil }
> 
> May looks more 'Ruby-ish' but isn't syntactically correct Ruby (just tried
> it to be sure). [image: :wink:]
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/opf/openproject/pull/2239/files#r20983266.

Commit:
Use path helper for templated paths too

Commit:
Use more Ruby-ish value assignment

myabc commented:
@hschink I'm getting a 500 status code, where (I believe) I should get a 409: `{"error":"Couldn't update the resource because of conflicting modifications."}` / @NobodysNightmare

NobodysNightmare commented:
@myabc You did not specify what you did, but a 409 should occur when you decrement the `lockVersion` that you receive...

generally speaking **changing** the `lockVersion` should result in 409 as long as you do not change it to the current value on the server...

myabc commented:
@NobodysNightmare please ignore – @hschink pointed out an issue with Rails+Grape in development that results in the wrong status code. _Should_ work in production :-)

myabc commented:
:+1:
